src/mlpack/methods/ann/layer/instance_norm_impl.hpp

```diff
@@ -0,0 +1,143 @@
+/**
+ * @file methods/ann/layer/instance_norm_impl.hpp
+ * @author Anjishnu Mukherjee
+ *
+ * Implementation of the Instance Normalization Layer.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_INSTANCENORM_IMPL_HPP
+#define MLPACK_METHODS_ANN_LAYER_INSTANCENORM_IMPL_HPP
+
+// In case it is not included.
+#include "instance_norm.hpp"
+
+namespace mlpack {
+namespace ann { /** Artificial Neural Network. */
+
+template<typename InputDataType, typename OutputDataType>
+InstanceNorm<InputDataType, OutputDataType>::InstanceNorm() :
+    size(0),
+    eps(1e-8),
+    average(true),
+    momentum(0.0),
+    deterministic(false),
+    reset(false)
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+InstanceNorm<InputDataType, OutputDataType>::InstanceNorm(
+    const size_t size,
+    const size_t numFunctions,
+    const double eps,
+    const bool average,
+    const double momentum) :
+    size(size),
+    numFunctions(numFunctions),
+    eps(eps),
+    average(average),
+    momentum(momentum),
+    deterministic(false),
+    reset(false)
+{
+    batchNorm = ann::BatchNorm<> (size * numFunctions, eps, average, momentum);
+    runningMean.zeros(size, 1);
+    runningVariance.ones(size, 1);
+    runningVariance = batchNorm.TrainingVariance();
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void InstanceNorm<InputDataType, OutputDataType>::Forward(
+    const arma::Mat<eT>& input,
+    arma::Mat<eT>& output)
+{
+  // Instance Norm with (N, C, H, W) is same as Batch Norm with (1, N*C, H, W),
+  // where N is the batchSize, C is the number of channels, H and W are the
+  // height and width of each image respectively.
+  if (!reset)
+  {
+    batchNorm.Reset();
+    reset = true;
+  }
+
+  shapeA = input.n_rows;
+  shapeB = input.n_cols;
+
+  if (deterministic)
+    batchNorm.Deterministic() = true;
+
+  arma::mat inputCopy(const_cast<arma::Mat<eT>&>(input).memptr(), shapeA*shapeB,
```

Comment:
I might be wrong, but I think the batchNorm layer handles this already, so we could pass the input directly.

Reply:
Instance Norm applied on a (N,C,H,W) input is equivalent to BatchNorm applied on a (1,N*C,H,W) input. (discussed in a comment above in more detail)
This is the conversion that this line tries to achieve. 
Not sure if the input passed to the BatchNorm directly would give the same results. (probably not I guess)